### PR TITLE
Trigger produt_updated webhook for collections and categories

### PR DESCRIPTION
**Pull request by d-wysocki**
I want to merge this change because...I want to trigger `PRODUCT_UPDATED` webhook for

- `CollectionCreate`
- `CollectionDelete`
- `CollectionAddProducts`
- `CollectionRemoveProducts`
- `CollectionBulkDelete`
- `CategoryDelete`
- `CategoryBulkDelete`

# Impact

* [ ] New migrations
* [ ] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [ ] Documentation needs to be updated

# Pull Request Checklist

* [ ] Privileged queries and mutations are guarded by proper permission checks
* [ ] Database queries are optimized and the number of queries is constant
* [ ] Database migration files are up to date
* [ ] The changes are tested
* [ ] GraphQL schema and type definitions are up to date
* [ ] Changes are mentioned in the changelog

### CODE REVIEW

**Comment by patrys** [comment]
Task linked: [SALEOR-2245 Removing and adding product in collections/categories does not trigger PRODUCT_UPDATE webhook](https://app.clickup.com/t/2549495/SALEOR-2245)

File: saleor/graphql/product/bulk_mutations/products.py
Comment (korycins) [inline saleor/graphql/product/bulk_mutations/products.py:98]: IMO, you need distinct here.
image

File: saleor/graphql/product/mutations/products.py
Comment (korycins) [inline saleor/graphql/product/mutations/products.py:260]: Do you need this empty method?

File: saleor/graphql/product/mutations/products.py
Comment (korycins) [inline saleor/graphql/product/mutations/products.py:286]: Do you need to convert them to list?
What about prefetches?  variant, categories, collections,  attributes(?)

**Review by korycins — COMMENTED** [review]
IMO Missing prefetches can cause a huge amount of DB queries here.

File: saleor/graphql/product/mutations/products.py
Comment (fowczarek) [inline saleor/graphql/product/mutations/products.py:233]: Should we prefetch any data?

File: saleor/graphql/product/mutations/products.py
Comment (fowczarek) [inline saleor/graphql/product/mutations/products.py:415]: Should we make queryset with prefetched needed data from `products` list?

File: saleor/graphql/product/mutations/products.py
Comment (fowczarek) [inline saleor/graphql/product/mutations/products.py:464]: Should we make queryset with prefetched needed data from products list?

File: saleor/product/utils/__init__.py
Comment (fowczarek) [inline saleor/product/utils/__init__.py:54]: Prefetch?

File: saleor/graphql/product/bulk_mutations/products.py
Comment (maarcingebala) [inline saleor/graphql/product/bulk_mutations/products.py:56]: Why do we override `perform_mutation` here? To pass the manager as an extra argument? It is available in the `info` object anyway.

Comment (d-wysocki) [reply saleor/graphql/product/bulk_mutations/products.py:56]: I override this method because, I wanted to pass manager to `bulk_action` method.
`perform_mutation` method in BaseBulkMutation takes `**data` parameter and pass it to `bulk_action`. So I added managet to `data` argument.

File: saleor/graphql/product/mutations/products.py
Comment (d-wysocki) [reply saleor/graphql/product/mutations/products.py:286]: I converted this into list because, this is delete mutation. Without casting to a list, after deleting collection, `instance` will doesn't had `products`.

I will add prefetch.

File: saleor/graphql/product/mutations/products.py
Comment (d-wysocki) [reply saleor/graphql/product/mutations/products.py:233]: Yes, I will add prefetch.

File: saleor/graphql/product/mutations/products.py
Comment (d-wysocki) [reply saleor/graphql/product/mutations/products.py:415]: Yes, I will add prefetch.

File: saleor/graphql/product/bulk_mutations/products.py
Comment (korycins) [reply saleor/graphql/product/bulk_mutations/products.py:56]: So maybe we should extend bulk_action to handle info by default? All future bulk mutations which use manager will require the same action,

File: saleor/graphql/product/mutations/products.py
Comment (d-wysocki) [reply saleor/graphql/product/mutations/products.py:260]: Yes, `CollectionUpdate` inherits from `CollectionCreate`, so I needed to clear this method to avoid triggering `PRODUCT_UPDATED` webhook when collection is updated.

File: saleor/graphql/product/mutations/products.py
Comment (korycins) [reply saleor/graphql/product/mutations/products.py:286]: > I converted this into list because, this is delete mutation. Without casting to a list, after deleting collection, `instance` will doesn't had `products`.

My bad, I missed the name of the mutation.

File: saleor/graphql/product/bulk_mutations/products.py
Comment (maarcingebala) [reply saleor/graphql/product/bulk_mutations/products.py:56]: Yep, was going to write the same as what Maciek suggests.

Comment (d-wysocki) [reply saleor/graphql/product/bulk_mutations/products.py:56]: Ok, I will change that.

File: saleor/graphql/product/mutations/products.py
Comment (korycins) [reply saleor/graphql/product/mutations/products.py:260]: So in that case I would call the product_updated inside a CollectionCreate.perform_mutation
or at least add the comment why we have empty method here <- which imho is not the best way.
@fowczarek  What do you think?

Comment (d-wysocki) [reply saleor/graphql/product/mutations/products.py:260]: Please correct me if im wrong, but `CollectionUpdate` also inherits `perform_mutation` from `CollectionCreate`.

File: saleor/graphql/product/bulk_mutations/products.py
Comment (IKarbowiak) [reply saleor/graphql/product/bulk_mutations/products.py:56]: I already added info to core `bulk_action` in #7052, it looks like we can merge it, so the problem will be solved.

File: saleor/graphql/product/mutations/products.py
Comment (d-wysocki) [reply saleor/graphql/product/mutations/products.py:484]: I added prefetch_related.

File: saleor/product/utils/__init__.py
Comment (d-wysocki) [reply saleor/product/utils/__init__.py:56]: I added prefetch_related.

File: saleor/graphql/product/mutations/products.py
Comment (korycins) [reply saleor/graphql/product/mutations/products.py:262]: Oh, yes you're right. I missed thath!

File: saleor/graphql/product/mutations/products.py
Comment (korycins) [inline saleor/graphql/product/mutations/products.py:416]: Please double check the `attributes` prefetch. I think that it misses some relations.

Comment (d-wysocki) [reply saleor/graphql/product/mutations/products.py:416]: Good catch, thanks. I will go thought all product_updated and product_variant_updated to add prefetch.

File: saleor/graphql/meta/mutations.py
Comment (korycins) [inline saleor/graphql/meta/mutations.py:122]: Why you can't call this prefetch inside the function responsible for Product extra actions?

File: saleor/graphql/product/mutations/products.py
Comment (korycins) [reply saleor/graphql/product/mutations/products.py:260]: But at least comment why we have empty method here

File: saleor/graphql/product/mutations/products.py
Comment (korycins) [inline saleor/graphql/product/mutations/products.py:286]: I don't think that you need `.all()` here.

File: saleor/product/models.py
Comment (korycins) [inline saleor/product/models.py:302]: IMO `prefetched_for_webhook` has enough information. No need to duplicate the `product` word
`Product.objects. prefetched_for_webhook()`

File: saleor/product/models.py
Comment (korycins) [inline saleor/product/models.py:307]: `variants__stocks` is not needed here.

File: saleor/product/models.py
Comment (korycins) [inline saleor/product/models.py:420]: Would be good to keep some consistency here. You have `prefetched_for_webhook` and here `prefetch_variant_updated`. I don't think that everyone will connect that this method is prefetch for webhook with name  PRODUCT_VARIANT_UPDATED`

File: saleor/product/utils/__init__.py
Comment (korycins) [inline saleor/product/utils/__init__.py:47]: Are you aware that you're applying prefetches on the qs which is empty?

File: saleor/graphql/product/mutations/products.py
Comment (fowczarek) [inline saleor/graphql/product/mutations/products.py:700]: We should add this prefetch before the merge.

File: saleor/product/models.py
Comment (fowczarek) [inline saleor/product/models.py:305]: Are you sure that's all the required fields for `generate_product_payload`?

Comment (d-wysocki) [reply saleor/product/models.py:310]: For single object we need only prefetch `attributes`, when we iterating over object then we need to prefetch everything related to that object.

**Comment by maarcingebala** [comment]
From my perspective, the code looks good, but the number of DB queries that this PR adds (according to the benchmark) worries me. Are these expected results?

**Review by maarcingebala — APPROVED** [review]

**Review by IKarbowiak — APPROVED** [review]

**Comment by maarcingebala** [comment]
@fowczarek @korycins This PR is waiting for your final review, can you take a look?

**Review by fowczarek — APPROVED** [review]

File: saleor/product/models.py
Comment (korycins) [inline saleor/product/models.py:427]: Just to confirm: What about `media`? I see that payload generated for variant uses it.

File: saleor/graphql/product/tests/test_collection.py
Comment (korycins) [inline saleor/graphql/product/tests/test_collection.py:1044]: The name suggests that the value here is 0.  Not sure that it is correct as, in line 1051, we have subtraction from this value.

File: saleor/product/models.py
Comment (d-wysocki) [reply saleor/product/models.py:427]: Yes, I tested it.

Comment (korycins) [reply saleor/product/models.py:427]: We have N queries here, where N is the number of media assigned to variant. 
First case: Variant without prefetch and some media objects. For each media obj, it generates one query.
Second: Prefetch with const number of queries
image

Comment (d-wysocki) [reply saleor/product/models.py:427]: Fixed.

**Review by korycins — APPROVED** [review]

**Pull request merged**
